---
title: "R_Assignment_Workflow_Updated"
author: "Sean McLaughlin"
date: "3/23/2021"
output: html_document
---

## R_Assignment: Part 1

## Data Inspection

# Loading Packages

```{r}
library(tidyverse)
library(ggplot2)
library(dplyr)
library(plyr)
library(reshape2)

# Various packages for data manioulation and display such as reshape2 and ggplot
```

# Loading Data to be Analyzed

```{r}
fang.data <- read.delim("https://raw.githubusercontent.com/SeanMcLaughlin-TT/BCB546X_RAssignment_SM/main/Data/fang_et_al_genotypes.txt", header=TRUE, row.names=NULL)
View(fang_et_al_genotypes)
snp.data <- read.delim("https://raw.githubusercontent.com/SeanMcLaughlin-TT/BCB546X_RAssignment_SM/main/Data/snp_position.txt", header=TRUE, row.names=NULL)
View(snp_position)
```

# Data Analysis

To analyze the data we run various commands that will tell us various aspects of the data to be manipulated such as data structure, type, names, rows, columns, and the files size in bytes.

```{r}
typeof(fang.data) #Indicates data structure type
dim(fang.data) #Shows row and columns number
sapply(fang.data, class) #Indicates data class type
names(fang.data) #Gives names of list
str(fang.data) #Shows data structure
file.info(list.files(path = "https://raw.githubusercontent.com/SeanMcLaughlin-TT/BCB546X_RAssignment_SM/main/Data/fang_et_al_genotypes.txt", full.names = TRUE)) #Gives file size
```


```{r}
typeof(snp.data) #Indicates data structure type
dim(snp.data) #Shows row and columns number
sapply(snp.data, class) #Indicates data class type
names(snp.data) #gives names of list
str(snp.data) #Shows data structure
file.info(list.files(path = "https://raw.githubusercontent.com/SeanMcLaughlin-TT/BCB546X_RAssignment_SM/main/Data/snp_position.txt", full.names = TRUE)) #Gives file size
```

## Data Processing

Our assignment calls for data of only Maize and Teosinte from fang_et_al so here we separate the data

# Separating Data into Maize and Teosinte
# Formatting relevant columns in SNP data
```{r}
Maize.Grouped <- filter(fang.data, Group == "ZMMIL"|Group == "ZMMLR"|Group =="ZMMMR")
Teosinte.Grouped <- filter(fang.data, Group == "ZMPBA"|Group == "ZMPIL"|Group =="ZMPJA")
snp.formated <- snp.data[c(1,3,4)] #Removing unnecessary columns
```

To combine the data we have to transpose our data and merge 
based on the SNP_ID data and also get rid of the unknown and multiple characters in the snp data by only taking the numeric data.

# Transposing and Formatting Data

```{r}
Maize.trans <- as.data.frame(t(Maize.Grouped)) # Transposing data
Teosinte.trans <- as.data.frame(t(Teosinte.Grouped))

chromosomes <- c(1,2,3,4,5,6,7,8,9,10) # removing unknown and multiple in chromosomes column
snp.edited <- subset(snp.formated, Chromosome %in% chromosomes)
```

# Merging Data

Now we merge our two datasets for both Maize and Teosinte

```{r}
Maize.merge <- merge(Maize.trans,snp.edited, by.x = "row.names", by.y = "SNP_ID")
Teosinte.merge <- merge(Teosinte.trans,snp.edited, by.x = "row.names", by.y = "SNP_ID")

Maize.merge[Maize.merge =="NA"] <- "?"
Teosinte.merge[Teosinte.merge =="NA"] <- "?"
```

Now that we have our merged data we need to sort it so that we have both ascending and descending with both Maize and Teosinte while replacing ? with other objects

# Data sorting

```{r}
Maize.asc <- arrange(Maize.merge, Position) #Sort ascending
Teosinte.asc <- arrange(Teosinte.merge, Position)

Maize.dec <- arrange(Maize.merge, desc(Position)) # Sort descending
Teosinte.dec <- arrange(Teosinte.merge, desc(Position))

```

We will now use a recursive code that will output a .csv file with our assignment outputs for each chromosome and species type.

# Ascending Outputs

``` {r}
 for(i in 1:10){
   ChrM.asc.num <- Maize.asc[Maize.asc$Chromosome == i]
  write.csv(ChrM.asc.num, paste0("Maize_chr_", i ,"_ascending.csv"), row.names = FALSE)}

 for(i in 1:10){
   ChrT.asc.num <- Teosinte.asc[Teosinte.asc$Chromosome == i]
  write.csv(ChrT.asc.num, paste0("Teosinte_chr_", i ,"_ascending.csv"), row.names = FALSE)}
```

# Decending Outputs

```{r}
 for(i in 1:10){
   ChrM.dec.num <- Maize.dec[Maize.dec$Chromosome == i]
  write.csv(ChrM.dec.num, paste0("Maize_chr_", i ,"_decending.csv"), row.names = FALSE)}

 for(i in 1:10){
   ChrT.dec.num <- Teosinte.dec[Teosinte.dec$Chromosome == i]
  write.csv(ChrT.dec.num, paste0("Teosinte_chr_", i ,"_decending.csv"), row.names = FALSE)} 
  
```

This should output the expected files 

###Part II

First we transpose the fang_et_al data and then merge with snp_positions again without separating the data

#Pivot_longer

```{r}
Fang.trans <- as.data.frame(t(fang.data))
Fang.Merged.gg <- merge(Fang.trans, snp.data, by.x = "row.names", by.y = "SNP_ID")

```

##SNPs per Chromosome and distribution

```{r}
Fang.Merged.gg$Chromosome <- factor(Fang.Merged.gg$Chromosome, levels = c("1","2","3","4","5","6","7","8","9","10"))
ggplot(data = Fang.Merged.gg) + geom_bar(aes(Fang.Merged.gg$Chromosome)) + xlab("Chromosome Number") + ylab("SNPs") # Shows SNP number by chromosome by factoring the chromosomes individually and others

ggplot(data = fang.data, aes(Group)) + geom_bar() + xlab("Groups") + ylab("SNPs") #Shows distribution of SNPs by different groups using fang.data data

```
##Missing data and amount of heterozygousity

```{r}
Names <- colnames(fang.data)[-c(1:3)]
fang.melt <- melt(fang.data, measure.vars = Names)
colnames(fang.melt)[c(3,4,5)] <- c("Group","SNP_ID", "Allele") #Melting data for clarity and for to further work it

fang.melt$Ho <- (fang.melt$Allele =="A/A" | fang.melt$Allele =="C/C" | fang.melt$Allele =="G/G" | fang.melt$Allele =="T/T")
Fang.melt.sort <- arrange(fang.melt, Sample_ID, Group)

sum.ID <- ddply(fang.melt.sort, c("Sample_ID"), summarise, total_ho = sum (Ho, na.rm=TRUE), total_het = sum (!Ho, na.rm=TRUE), missing = sum(is.na(Ho)))# Setting parameters to show ratio of missing data and heterozygous genotypes

sum.melt <- melt(sum.ID, measure.vars = c("total_ho", "total_het", "missing"))

ggplot(sum.melt,aes(x = Sample_ID, y = value, fill=variable)) + geom_bar(stat = "identity", position = "stack")
attributes(sum.melt)
```

##Own Visualization

I was wondering what the JG_OTU column was, so I ploted the names in that column against the associated SNPs. The result I got was expected as every row has a corresponding JG_OTU name.

```{r}
ggplot(data = fang.data, aes(JG_OTU)) + geom_bar() + xlab("JG_OTU") + ylab("SNPs")

```



